proposals/list-patterns.md

```diff
@@ -121,6 +172,141 @@ expr.Length is >= 2
 ```
 The *input type* for the *slice_pattern* is the return type of the underlying `this[Range]` or `Slice` method with two exceptions: For `string` and arrays, `string.Substring` and `RuntimeHelpers.GetSubArray` will be used, respectively.
 
+#### Lowering on enumerable type
+
+> **Open question**: Confirm that async enumerables are out-of-scope.  
+> **Open question**: Confirm that slice patterns with a sub_pattern (such as `..var x`) are out-of-scope.  
+
+Although a helper type is not necessary, it helps simplify and illustrate the logic.
```

Comment:
I gather the following is not the rigorous codegen spec and we may make changes as we go through the impl?

Reply:
Right.

This rough codegen was mostly to convince myself that it would fit in the DAG binding design (evaluation steps, test steps, ...). Overall I expect it will fit okay.

The parts I'm not sure yet:
- how do we account for evaluation steps returning a value or not? (in contrast, property accesses always succeed)
- who is responsible for the end-of-enumeration count check (when exiting the list-pattern, like `{ 1, 2 }`)?
- can we offer enough guarantees in the DAG binding that we don't need to cache starting values at all? (those values would get cached into temps if they are enumeratoed in the right order)
- is there enough helper logic that we could extract to a BCL type?
- how do we keep track of enumerators to dispose?

Reply:
>how do we account for evaluation steps returning a value or not?

We either use an evaluation followed by a standard test, or a dedicated test node. As long as these won't affect subsumption checking, we can do away with simple eval nodes. In the prototype, I tried the first approach initially, but with trailing patterns, had to introduce a test node to disallow MoveNext after a slice which can happen in multi-arm matches.

>who is responsible for the end-of-enumeration count check 

If the enumerator has a length (e.g. via TryGetNonEnumeratedCount), we check that first for an early failure, otherwise we only count if there's a length pattern. This happens right after we enumerate the sequence to the end. While we're doing it, we don't want to exceed the maxLength (if specified) and after that, we want to ensure we've reached the minLength (either inferred or specified). This cover the "early failure" that was discussed in the last LDM.

Note that a simple `helper.Count() is P` always enumerate the sequence to the end which is not what we want, therefore we need to compute the actual value set that we're testing for. Since we operate on a range of values, something like `[1 or 3]` causes incorrect results and should be gated.

>can we offer enough guarantees in the DAG binding that we don't need to cache starting values at all? (those values would get cached into temps if they are enumerated in the right order)

That is correct. Starting patterns won't need any kind of buffer since we have a temp per each.

>is there enough helper logic that we could extract to a BCL type?

I think if we use a standard buffer type we can indeed propose to it BCL. Currently I used a generic fixed-size stack with no additional logic.


>how do we keep track of enumerators to dispose?

Since we can't "join" leaf nodes in a DAG, we just wrap the "rest" of the lowering in a try/finally, starting at each GetEnumerator. As a consequence enumerators might stack up. For instance,  in `{ enum1: {0}, enum2: {0} }`, both enumerators won't get disposed until after the whole pattern is executed.